### ISSUE TITLE
Using a custom palette made through mata interpolated colors

### ISSUE THREAD

**Issue by ericmelse** [benjann/colrspace#1]
Dear Ben,
Your work on color palettes is such an inspirtation.
What I have done is to design a revised color (coding) scale based on CET CBD1 from Kovesi ([2015](https://colorcet.com/)):
![Palette_CET_CBD1](https://github.com/benjann/colrspace/assets/3686136/30ce0693-f723-4d53-b0b5-e1a94c4e6290)
My objective is to interpolated the start and end point colors of this scale but shift the neutral position to the left, using your mata code example, like:
```
. mata: A = ColrSpace()
. mata: A.colors("#3A90FE #EDEDED #A89008")
. mata: B = C = J(1, 1, A)
. mata: A.ipolate(100) //
. mata: B.ipolate(100, "", ., ., (0, .15, 1)) // (shift left)
. colorpalette, gropts(ysize(1.5) scale(*2.67)) nonumbers: mata(A) / mata(B)
```
which results in:
![Palette_CBD1](https://github.com/benjann/colrspace/assets/3686136/23c3b365-a5a3-4eb1-886f-5dfbbeb01cb1)
My next step is to use (B), the 'shifted scale of colors', using `heatplot`, after using the `colorpalette` command, like:
`heatplot D, aspectratio(1) values(format(%9.2f)) color(mata(B), intensity(.6)) xsize(6) ysize(5) lower nodiagonal`
 which results in:
![Model_nlsw88_CBD1](https://github.com/benjann/colrspace/assets/3686136/f955c189-028d-481a-8cb6-2d8594353406)
I suppose that because of using 100 steps in the color scale the 'legend' is now somewhat 'disturbed'. Assuming that it is not a regular legend but created by your ado (as a 'ramp') with a title, it now is an issue for me how to:
- disable the display of the legend/ramp title (here it is the name of the matrix **D**)?
- control the color step labels?

Because of the current display of the legend/ramp title, the addition of the legend part forces the x-axis of the heatplot downwards, hence my wish to remove the title (before I modify other graph elements to adjust).
Next, I would like to control the number of the color step labels and their format (typically rounded to two decimals with a .05 step size).

So, maybe you can advise me how to proceed.

**Comment by benjann** [comment]
Note that you could get the palette as

```
colorpalette CET CBD1, n(3) ipolate(100, positions(0 .15 1))
```

or directly in `heatplot` as

```
heatplot ..., color(CET CBD1, n(3) ipolate(10, positions(0 .15 1)) intensity(.6)) ...
```

No need to go through Mata. The title of the legend you can omit by adding option `legend(subtitle(""))`. However, this would not fix the spacing issue with the xaxis. The displayed legend in your graph is a standard Stata legend even if it looks like a ramp. `heatplot` automatically prunes some of the labels if there are too many; see option `keylabels()`. Furthermore, Stata just cannot squeeze the legend enough because it has so many keys (100), and the effect is that the xaxis gets pushed down a bit; such odd behavior sometimes happens if a legend consumes too much space. Anyhow, using a standard legend if you have 100 colors is not really feasible. I would suggest using the `ramp()` option instead, e.g. as folows:

```
mat D = J(9,9,0)
mat D[2,1] = .69
mat D[1,2] = .69
mat D[5,3] = 1
mat D[3,5] = 1
mat D[5,4] = .13
mat D[4,5] = .13
heatplot D, aspectratio(1) values(format(%9.2f)) lower nodiagonal xsize(6) ysize(5) ///
    color(CET CBD1, n(3) ipolate(100, positions(0 .15 1)) intensity(.6)) ///
    ramp(right space(12) subtitle("") ylabel(0(.1)1, labsize(small)))
```

ben

**Comment by ericmelse** [comment]
Dear Ben,
Thank you for your helpful instruction. I have proceeded with some further improvement using `p()` and `ramp` options, like:
```
heatplot D, aspectratio(1) values(format(%9.2f)) lower nodiagonal xsize(6) ysize(5) ///
    color(CET CBD1, n(3) ipolate(100, positions(0 .15 1)) intensity(.6)) p(lc(white) lw(*.1))  ///
    ramp(right space(12) length(94) graphreg(m(t-1)) subtit("P-value", s(small) pos(6) m(t+1.5)) ///
	ylab(0(.05)1, labsize(vsmall)) ysca(noli))
```
Which results in:
![Model_nlsw88_CBD1_ramp](https://github.com/benjann/colrspace/assets/3686136/e992ab81-13b1-4cc9-9d74-082720b5ec1a)
This is just fine, but, I also tried to control the options to visualize (or not) the ramp/legend tick marks, e.g. by including the option `ytick(, noticks)) `, like:
```
heatplot D, aspectratio(1) values(format(%9.2f)) lower nodiagonal xsize(6) ysize(5) ///
    color(CET CBD1, n(3) ipolate(100, positions(0 .15 1)) intensity(.6)) p(lc(white) lw(*.1))  ///
    ramp(right space(12) length(94) graphreg(m(t-1)) subtit("P-value", s(small) pos(6) m(t+1.5)) ///
   ylab(0(.05)1, labsize(vsmall)) ysca(noli)) ytick(, noticks)) 
```
The issue now is that `heatplot` will produce the plot which is identical to the above plot, i.e. including the tick marks. No error message is reported. Also specifying tick mark properties like` ytick(, tlength(*.2) tlwidth(*.5))` will result in the same plot without an error message.
So, my question now is if I am not coding the tick mark options properly, or if that is not supported by `heatplot`, or if there is some internal issue here that requires your attention.

**Comment by benjann** [comment]
You need to include suboption `notick` in option `ylabel()` within `ramp()`.

**Comment by ericmelse** [comment]
Yes, indeed! By including `ylab(0(.05)1, labsize(vsmall) tl(*.6) tlw(*.5))`, the result is now:
![Model_nlsw88_CBD1_ramp_optimized](https://github.com/benjann/colrspace/assets/3686136/a5c9972f-12a1-4c81-954d-2e03367df4ba)

Note that I have improved the distinguishability of the color steps between [.15,1] by only manipulating the intensity of the first (leftmost) color using:
`color("58 144 254*.6" "237 237 237" "168 144 8", n(3) ipolate(100, positions(0 .15 1)))`
The reason is that from .15 down to 0 the 'statistical relevance' supposedly increases while from .15 up to 1 that decreases. Using the original color at the 1 position provides more distinguishable steps up to that point (in my opinion). Of course, using .15 as a midpoint is a personal choice. The reason why I reduced the intensity of the leftmost (blue) color is that so many of the coded squares in the plot have a value of about 0 and thus these dominate the visual impression of this plot. Hence my need to subdue to color blue.
Now, I inspect all these steps and their definition by `colorpalette`, which results in:
![Colorpalette_CET_CBD1_Optimized](https://github.com/benjann/colrspace/assets/3686136/8a685652-977d-4893-ad59-05f249976a51)
As such this is fine except for the labels of the first 15 colors. Because of the intensity manipulation, their multiplier value runs over the step indicator of the colors in the second column.
I am totally guilty of looking for perfection and probably ask for something close to irritation. But, maybe there is already an option to round such multipliers to 2 decimals, or, possibly that should be something that is better coded in the ado itself.

**Comment by ericmelse** [comment]
I just read your recommendation to use `intensify(.6)` and that is the solution to my request.

**Comment by benjann** [comment]
I am afraid there is no option to fix this. The graphs created by `colorpalette` are really just meant for quick inspection of colors, so I did not bother too much about making it flexible.

**Comment by ericmelse** [comment]
Yes, I agree and it is actually not really necessary because `intensify(#)` takes care of a 'clean' label. Most remarkable that you provide this option (and I should have read your help file more closely!).
